src/compiler/moduleNameResolver.ts

```diff
@@ -816,6 +818,33 @@ namespace ts {
         return loadNodeModuleFromDirectory(extensions, candidate, failedLookupLocations, onlyRecordFailures, state, considerPackageJson);
     }
 
+    function parseNodeModuleFromPath(path: string): { packageDirectory: string, subModuleName: string } | undefined {
+        const nodeModules = "/node_modules/";
+        const idx = path.indexOf(nodeModules);
+        if (idx === -1) {
+            return undefined;
+        }
+
+        const indexAfterNodeModules = idx + nodeModules.length;
+        let indexAfterPackageName = indexOfNextSlash(path, indexAfterNodeModules);
+        if (path.charCodeAt(indexAfterNodeModules) === CharacterCodes.at) {
+            indexAfterPackageName = indexOfNextSlash(path, indexAfterPackageName);
+        }
+        const packageDirectory = path.slice(0, indexAfterPackageName);
+        const subModuleName = removeExtensionAndIndex(path.slice(indexAfterPackageName));
+        return { packageDirectory, subModuleName };
+    }
+
+    function indexOfNextSlash(s: string, pos: number): number {
+        const x = s.indexOf(directorySeparator, pos);
+        return x === -1 ? pos : x;

```

Comment:
This special case makes sense for this application, but probably not in general.  Would it make sense to move this helper into `parseNodeModuleFromPath`?

Reply:
I'm not a fan of moving things into closures if they don't need to close over any variables, but this could be named better.